### DISCUSSION TITLE
Infer Type Argument of Generic Return Type by Examining Parameter Generic Types

### DISCUSSION THREAD

**Discussion by wessleym** [dotnet/csharplang#7262] [Language Ideas]
I have a C# method that is requiring explicit type arguments, and I am getting this error:

`error CS0411: The type arguments for method 'Test.TargetMethod<TObjectFactory, TResult>(TObjectFactory)' cannot be inferred from the usage. Try specifying the type arguments explicitly.`

Please see the below example code, specifically the two calls to `TargetMethod` near the bottom. In the first invocation of `TargetMethod`, the type arguments should be obvious since I'm passing in a `StringFactory`, which only returns a `string`.  (See the generic type constraints of `TargetMethod`.) Could C# be made to realize that it has all it needs to know and shouldn't need type arguments? It should know the return type based on the type of the first parameter. Thank you.

```
public interface IObjectFactory<out TResult>
{
    TResult Construct();
}

public class StringFactory : IObjectFactory<string>
{
    public string Construct() { return ""; }
}

public static class Test
{
    //I need to be able to pass in a generic factory to this method.
    //I also need it to return a generic result.
    //This is a mandatory part of the design, and on its own, it works fine.
    private static TResult TargetMethod<TObjectFactory, TResult>(TObjectFactory objectFactory)
        where TObjectFactory : IObjectFactory<TResult>
    {
        return objectFactory.Construct();
    }

    //This method can't easily call the other one.
    public static void EntryMethod()
    {
        StringFactory stringFactory = new StringFactory();
        //On the below line, how can I get C# to figure out that TargetMethod will return a string?
        string result1 = TargetMethod(stringFactory);//compilation error
        //I can do the below, but I don't want to have to specify type arguments each time.
        string result2 = TargetMethod<StringFactory, string>(stringFactory);
    }
}
```

**Comment by CyrusNajmabadi** [comment]
This is currently by design. The second type parameter isn't in the signature (except in constraints), and the language doesn't use constraints when doing inference.

**Reply by RikkiGibson** [reply]
Some issue or discussion definitely exists already proposing to use constraints in type inference. It results in breaking changes. Not to say that those breaking changes mean "100% it will never happen", but that it definitely increases the risk and complexity around doing it.

**Reply by 333fred** [reply]
There are 2 related issues:

1. Inferring types from constraints. That's https://github.com/dotnet/csharplang/discussions/741.
2. The actual solution you want for this specific case, associated types. That's https://github.com/dotnet/csharplang/issues/5556.

**Comment by wessleym** [comment]
@CyrusNajmabadi, yes, I am aware. Thank you. But I'm suggesting it would be beneficial for that behavior to change.
@RikkiGibson, it's hard for me to imagine how this could be a breaking change. It would only allow new code to compile. Any existing code would have already specified type arguments and would be unaffected. The change would allow the type arguments to be omitted.
@333fred, I agree that #741 should be helpful. #5556 seems beyond what I am proposing.

**Reply by 333fred** [reply]
Associated types are what you actually want here (and of the two I linked, they're the one we're interested in). IObjectFactory really doesn't need to be externally generic: it's a detriment in cases like you've shown here. For any given implementation, there's only one valid choice, and it was made when the type was implemented.

**Reply by wessleym** [reply]
I think I see now. Thank you.
